ipam/ippool_manager.go

```diff
@@ -400,41 +409,54 @@ func (m *IPPoolManager) createAddress(ctx context.Context,
 	return addresses, nil
 }
 
-// DeleteDatas deletes old secrets
+// deleteAddress removes the finalizer from the IPClaim and deletes the associated IPAddress
 func (m *IPPoolManager) deleteAddress(ctx context.Context,
 	addressClaim *ipamv1.IPClaim, addresses map[ipamv1.IPAddressStr]string,
 ) (map[ipamv1.IPAddressStr]string, error) {
 
-	m.Log.Info("Deleting Claim", "IPClaim", addressClaim.Name)
+	m.Log.Info("Deleting IPAddress associated with IPClaim", "IPClaim", addressClaim.Name)
 
 	allocatedAddress, ok := m.IPPool.Status.Allocations[addressClaim.Name]
 	if ok {
 		// Try to get the IPAddress. if it succeeds, delete it
-		tmpM3Data := &ipamv1.IPAddress{}
+		ipAddress := &ipamv1.IPAddress{}
 		key := client.ObjectKey{
 			Name:      m.formatAddressName(allocatedAddress),
 			Namespace: m.IPPool.Namespace,
 		}
-		err := m.client.Get(ctx, key, tmpM3Data)
+		err := m.client.Get(ctx, key, ipAddress)
 		if err != nil && !apierrors.IsNotFound(err) {
 			addressClaim.Status.ErrorMessage = pointer.StringPtr("Failed to get associated IPAddress object")
 			return addresses, err
 		} else if err == nil {
-			// Delete the secret with metadata
-			err = deleteObject(m.client, ctx, tmpM3Data)
+			// Remove the finalizer
+			ipAddress.Finalizers = Filter(ipAddress.Finalizers,
+				ipamv1.IPAddressFinalizer,
+			)
+			err = updateObject(m.client, ctx, ipAddress)
+			if err != nil && !apierrors.IsNotFound(err) {
+				m.Log.Info("Unable to remove finalizer from IPAddress", "IPAddress", ipAddress.Name)
+				return addresses, err
+			}
+			// Delete the IPAddress
+			err = deleteObject(m.client, ctx, ipAddress)
 			if err != nil {
 				addressClaim.Status.ErrorMessage = pointer.StringPtr("Failed to delete associated IPAddress object")
 				return addresses, err
 			}
+			m.Log.Info("Deleted IPAddress", "IPAddress", ipAddress.Name)
```

Comment:
shall this line be added after line 463 ?

Reply:
Not really. In 442 we delete the IPAddress from Kubernetes, then in 463 we remove it from the allocations in the IPPool.

Reply:
Ah ok, I though by this you mean to remove it from the allocations in the IPPool, a bit confusing though

Reply:
Any suggestion how to make it clearer?

Reply:
this looks fine to me, we could improve/add a log to IP deletion in  allocations

Reply:
Thanks, I can do a followup with this :)